core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -156,10 +156,10 @@ class ReplicaAlterLogDirsThread(name: String,
    * @param partitions map of topic partition -> leader epoch of the future replica
    * @return map of topic partition -> end offset for a requested leader epoch
    */
-  def fetchEpochsFromLeader(partitions: Map[TopicPartition, Int]): Map[TopicPartition, EpochEndOffset] = {
+  def fetchEpochsFromLeader(partitions: Map[TopicPartition, Option[Int]]): Map[TopicPartition, EpochEndOffset] = {
     partitions.map { case (tp, epoch) =>
       try {
-        val (leaderEpoch, leaderOffset) = replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch)
+        val (leaderEpoch, leaderOffset) = replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch.getOrElse(UNDEFINED_EPOCH))
```

Comment:
It would be nice if we could limit use of `UNDEFINED_EPOCH` to the serde logic. I think passing around integer sentinels is unsafe because the type checker can't help us ensure that we have proper checking in place. A couple options come to mind:

1. We can let `endOffsetFor` accept `Option[Int]` as the epoch to query. The epoch if defined must be non-negative.
2. Keep the current `endOffsetFor` API, but require the epoch to be non-negative. We can match on the epoch here and only invoke `endOffsetFor` if we have an epoch. Otherwise we can return `(UNDEFINED_EPOCH, UNDEFINED_EPOCH_OFFSET` directly.

Reply:
Not sure which one is better, but I guess 2. might be better - the less `options` the simpler the code, right?
Do you think it is worth checking and raising an exception in `endOffsetFor` if a negative epoch is sent? I went without it